2단계 - 수강신청(도메인 모델)

ksh9891 opened the pull request:
안녕하세요 리뷰어님 :)
2단계 완료하여 PR 요청 드립니다.
리뷰 잘 부탁드립니다

ksh9891 committed:
doc(lms): 요구사항 문서화

ksh9891 committed:
doc(lms): 테스트 내용 문서화

ksh9891 committed:
feat(lms): Session 이 없을 때 Course 생성 시 예외 발생 처리

- Session 이 없는 경우 Course 생성 시 예외를 발생시키도록 구현
- 해당 동작을 검증하는 테스트 케이스 추가
- readme 파일에 테스트 완료 표시

ksh9891 committed:
feat(lms): 기수가 1보다 작은 수가 입력되었을 때Course 생성 시 예외 발생 처리

- 기수가 1 미만인 경우 Course 생성 시 예외를 발생시키도록 구현
- 해당 동작을 검증하는 테스트 케이스 추가
- readme 파일에 테스트 완료 표시

ksh9891 committed:
test(lms): 기수와 Session 이 포함되면 Course 가 정상적으로 생성되는 테스트 추가

- 기수와 Session 이 포함되면 Course 가 정상적으로 생성되는 테스트 추가
- readme 파일에 테스트 완료 표시

ksh9891 committed:
feat(lms): Session 생성 시, 시작일과 종료일 중 하나라도 존재하지 않으면 예외 발생처리

- Session 생성 시, 시작일과 종료일 중 하나라도 존재하지 않으면 예외 발생처리 로직 추가
- 테스트 케이스 추가
- readme 파일에 테스트 완료 표시

ksh9891 committed:
feat(lms): Session 생성 시, 종료일을 시작일 이전으로 설정을 하면 예외가 발생하도록 처리

- Session 생성 시, 종료일을 시작일 이전으로 설정을 하면 예외가 발생하도록 하는 로직 추가
- 테스트 케이스 추가
- readme 파일에 테스트 완료 표시

ksh9891 committed:
doc(lms): 오타수정

ksh9891 committed:
refactor(lms): 기간에 대해서 dateRange 객체를 생성하여 처리

- Session 에 존재하던 시작일, 종료일에 대한 유효성 검증 책임을 DateRange 에 위임
- 해당 테스트코드 이동
- 문서 수정

ksh9891 committed:
doc(lms): DateRange 테스트 케이스 추가

ksh9891 committed:
feat(lms): DateRange 객체에 isBetween 역할 추가

- 입력된 날짜가 기간 내의 날짜인지 확인하는 메서드 추가
- 테스트 케이스 추가
- 문서 수정 및 체크표시

ksh9891 committed:
feat(lms): Session 객체 생성 시 기간에 대한 값이 입력되지 않으면 예외가 발생하도록 추가

- 기간, 모집기간에 대한 예외처리 추가
- 테스트 케이스 추가
- 테스트 문서 체크

ksh9891 committed:
doc(lms): 불필요한 테스트 내용 제거

ksh9891 committed:
feat(lms): Session 이 모집중 상태가 아니면신청 시 예외가 발생하도록 처리

- Session 이 모집중 상태가 아니면신청 시 예외가 발생하는 로직 추가
- 테스트 코드 추가
- 테스트 문서 체크박스 체크

ksh9891 committed:
fix(lms): isRecruiting 메서드의 비교조건 버그 수정

ksh9891 committed:
feat(lms): Money 일급 객체 추가

- Money 일급 객체 추가
- 테스트 코드 추가
- 테스트 내용 추가 및 체크박스 체킹

ksh9891 committed:
refactor(lms): 누락된 final 키워드 추가

ksh9891 committed:
fix(lms): 오타수정

ksh9891 committed:
feat(lms): 수강생이 결제한 금액과 수강료가 일치하지 않으면 예외를 반환하는 기능 추가

- 수강신청 시 조건 추가
- Session 객체에 정적팩토리 메서드 추가
  - freeSession, paidSession
- 테스트 코드 추가
- 문서 수정

ksh9891 committed:
feat(lms): 수용인원에 대한 표현을 위한 일급객체 Capacity 추가

ksh9891 committed:
feat(lms): SessionManager 추가,
Session 에서 하던 수강신청의 역할을 SessionManager 이 하도록 변경

ksh9891 committed:
doc(lms): 테스트 명세 수정

ksh9891 committed:
feat(lms): 이미지 사이즈 관련 VO 객체 추가

ksh9891 committed:
feat(lms): 이미지 확장자 관련 Enum 추가

ksh9891 committed:
feat(lms): 이미지 properties 객체 생성

ksh9891 committed:
feat(lms): SessionCoverImage 객체 추가

csh0034 commented on STEP2-README.md:
```diff
@@ -0,0 +1,66 @@
+# 2단계 - 수강신청 (도메인)
```
요구사항 정리후 체크해나가시며 진행하신점 너무 좋네요 💯

csh0034 commented on src/main/java/nextstep/courses/domain/CourseV2.java:
```diff
@@ -0,0 +1,50 @@
+import java.time.LocalDateTime;
+import java.util.List;
+
+public class CourseV2 {
```
기존 코드를 수정해보시면 좋을것 같아요 🙂

csh0034 commented on src/main/java/nextstep/courses/domain/DateRange.java:
```diff
@@ -0,0 +1,31 @@
+    private final LocalDate startDate;
+    private final LocalDate endDate;
+
+    public DateRange(final LocalDate startDate, final LocalDate endDate) {
```
검증 잘해주셨네요 👍

csh0034 commented on src/main/java/nextstep/courses/domain/ImageProperties.java:
```diff
@@ -0,0 +1,27 @@
+package nextstep.courses.domain;
+
+public class ImageProperties {
```
List, Properties 등은 java, 프로그래밍 등에선 특정 타입등을 나타내기 때문에
혼동의 여지가 있을수있습니다. 적절한 이름을 고민해보시면 좋을것 같아요.

csh0034 commented on src/main/java/nextstep/courses/domain/SessionUser.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.Objects;
+
+public class SessionUser {
+    private final String userId;
```
유저의 경우 NsUser 를 사용해보시면 좋을것 같아요!

csh0034 commented on src/main/java/nextstep/courses/domain/Session.java:
```diff
@@ -0,0 +1,48 @@
+package nextstep.courses.domain;
```
session 관련 도메인의 경우 Course 와 생명주기가 다르므로 별도 패키지로 분리해보시는건 어떨까요?

csh0034 commented on src/main/java/nextstep/courses/domain/Session.java:
```diff
@@ -0,0 +1,48 @@
+import java.time.LocalDate;
+
+public class Session {
+    private final DateRange sessionDateRange;
```
`강의 상태는 준비중, 모집중, 종료 3가지 상태를 가진다.` , `강의 수강신청은 강의 상태가 모집중일 때만 가능하다.`

강의(Session) 이 상태를 가지면서 해당 상태에 따라 행위에 대한 검증이 필요하지 않을까요?

csh0034 commented on src/main/java/nextstep/courses/domain/SessionManager.java:
```diff
@@ -0,0 +1,55 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class SessionManager {
```
SessionManager 과 Session 의 차이가 모소한것 같아요.

보통 Service/Manager 등은 Service Layer 에서 사용하는 이름입니다.
분리가 필요할지 또는 적절한 이름에 대해서 고민해보시면 좋을것 같습니다.

추가로 Session 이 어떠한 Course 에 속하는지에 대한 정보가 필요하지 않을까요?

ksh9891 replied:
SessionManager 를 만들 당시에는 "Session 이 너무 많은 역할을 하고 있는 거 아닌가?" 라는 생각에 분리를 했었는데요, 피드백을 받고난 이후에 병합을 해보니 Session 에 역할을 부여해도 괜찮을 것 같아서, 병합하여 처리했습니다

ksh9891 replied:
현재 Session 에서는 Course 에 대한 정보가 필요하지 않아서 Course 만 Session 을 알고 있으면 될 것 같아 Session 필드로 추가하지 않았습니다 !

csh0034 commented on src/main/java/nextstep/courses/domain/SessionManager.java:
```diff
@@ -0,0 +1,55 @@
+        apply(applyDate, sessionUser, Money.ZERO);
+    }
+
+    public void apply(final LocalDate applyDate, final SessionUser sessionUser, final Money fee) {
```
`결제를 완료한 결제 정보는 payments 모듈을 통해 관리되며, 결제 정보는 Payment 객체에 담겨 반한된다.`

Payment 를 전달받아 처리해보는건 어떨까요?

추가로 요구사항의 경우 등록일을 검증하기보단 강의 상태 를 검증해야하므로 요구사항에 맞춰
변경해보시면 좋을것 같습니다.

ksh9891 replied:
상태를 기간으로 제어하고 있었는데, 별도의 상태 필드를 가져서 제어하는 방향으로 수정했습니다!

csh0034 reviewed (changes requested):
2단계 구현 잘해주셨네요 😄 
몇가지 코멘트 남겨두었는데 확인 부탁드려요!

ksh9891 committed:
refactor(lms): CourseV2 클래스 제거 및 Course 리팩토링

ksh9891 committed:
refactor(lms): 패키지 구조 수정

ksh9891 committed:
fix(lms): 클래스 이름 수정

ksh9891 committed:
fix(lms): sessionUser 를 NsUser 로 변경

ksh9891 committed:
doc(lms): 테스트 명세 수정

ksh9891 committed:
feat(lms): DateRange 역할 추가

- 입력값이 시작일 보다 이전인지 검증하는 역할
- 테스트 메서드 추가

ksh9891 committed:
feat(lms): Session 역할 추가 및 제거

- 강의기간과 모집기간을 별도로 제어하고 있었는데, 모집기간을 제거함
- 모집중 상태는 강의기간을 기준으로 제어되도록 수정
- 강의의 상태 추가 (준비중, 모집중, 종료)

ksh9891 committed:
refactor(lms): SessionManager 의 역할을 Session 이 담당하도록 변경

ksh9891 committed:
refactor(lms): SessionManager 의 역할을 Session 이 담당하도록 변경

ksh9891 committed:
refactor(lms): private 메서드로 변경됨에 따른 테스트 코드 제거

ksh9891 committed:
refactor(lms): private 메서드 위치 수정

ksh9891 committed:
feat(payment): Payment 객체에 금액검증 역할 추가

ksh9891 committed:
refactor(payment): Money 객체를 인자로 받던 부분을 Payment 객체로 전달받도록 변경

ksh9891 committed:
refactor(lms): Session 에서 상태를 필드로 가지고 있도록 변경

ksh9891 commented:
안녕하세요. 리뷰어님
피드백 주신 내용 잘 확인하였습니다.
고민해보고 수정하여 다시 PR 요청드립니다.
감사합니다 :)

csh0034 commented on src/main/java/nextstep/session/domain/SessionCoverImage.java:
```diff
@@ -0,0 +1,15 @@
+public class SessionCoverImage {
+    private final String name;
+    private final ImageExtension extension;
+    private final ImageDimensions properties;
```
필드명이 변경 안되었네요 😁

ksh9891 replied:
객체만 바꾸고 미쳐 수정하지 못했네요 ㅠ

csh0034 commented on src/main/java/nextstep/session/domain/Session.java:
```diff
@@ -0,0 +1,87 @@
+    private final SessionStatus sessionStatus;
+    private final Money fee;
+    private final Capacity capacity;
+    private final List<NsUser> sessionUserList;
```
Session 과 NsUser 는 다대다 입니다.

기존에 추가해주셨던 형태에서 userId 부분만 변경해보시는건 어떨까요?

csh0034 commented on src/main/java/nextstep/session/domain/SessionStatus.java:
```diff
@@ -0,0 +1,15 @@
+package nextstep.session.domain;
+
+public enum SessionStatus {
+    준비중(false), 모집중(true), 종료(false);
```
한글로 표현해주신 이유가 있으실까요? 🤔

ksh9891 replied:
업무에서 enum을 한글로 자주 사용하다 보니 자연스럽게 그렇게 작성하게 되었습니다. 
제 개인적인 생각으로는 영어보다 한글이 더 직관적이라는 느낌을 받아서 enum 타입을 한글로 선언하는 것도 괜찮다고 생각합니다.
혹시 한글로 enum을 사용하는 것에 대해 리뷰어님께서는 어떻게 생각하시는지 궁금합니다!

csh0034 replied:
컨벤션으로 한글을 사용하겠다고 정해놓았으면 문제 없다고 생각해요.
클래스명 또한 한글로 가능하지만 enum 내부에서만 사용하셔서 여쭤본 내용입니다.
일관성있게 작성하는것이 중요하다고 생각해요.

추가로 isRecruit 필드는 `this == 모집중` 을 통해 나타낼수 있으므로 제거 가능할것 같아요!

csh0034 commented on src/main/java/nextstep/courses/exception/SessionNotRecruitingException.java:
```diff
@@ -0,0 +1,7 @@
+package nextstep.courses.exception;
+
+public class SessionNotRecruitingException extends RuntimeException {
```
단순히 message 를 받는 대상이라면 RuntimeException 하위 표준 예외를 사용해볼수도 있지 않을까요?

추가로 위치도 같이 변경되어야겠네요!

ksh9891 replied:
> 단순히 message 를 받는 대상이라면 RuntimeException 하위 표준 예외를 사용해볼수도 있지 않을까요?  

SessionNotRecruitingException 과 같은 커스텀 예외를 던지기 보다는, RuntimeException 하위 표준 예외를 던지라고 말씀하신 것으로 이해했습니다.

해당 예외는 하위 표준 예외로 대체하고 커스텀예외는 제거하는 형식으로 처리했습니다

csh0034 commented on src/main/java/nextstep/session/domain/Session.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    public void apply(final NsUser sessionUser) {
+        apply(sessionUser, new Payment());
```
Payment 를 생성해야하는걸까요? 🤔

`강의는 무료 강의와 유료 강의로 나뉜다`

무료 강의의 경우엔 `capacity.isFull` 호출자체가 필요 없게 됩니다.
적절한 enum 타입을 통해 추가하여 구분해 볼 수 있지 않을까요?

ksh9891 replied:
강의료가 0원이 아닌 경우에 유료강의로 판단하여 유료강의에서만 검증해야 할 부분을 검증하도록 수정하였습니다

csh0034 replied:
물론 강의료가 0원 인지를 통해 무료/유료 강의를 구분할 수 있지만 요구사항의 경우 무료강의, 유료강의 일때 필요한 동작들이 있습니다.
현재 구현방식으로서 요구사항을 파악해본다면 하단과 같게 됩니다.

1. 강의료가 0원인 경우 최대 수강 인원 제한이 없다.
2. 강의료가 0원이 아닌 경우 강의 최대 수강 인원을 초과할 수 없다

요구사항을 보고 구현된 코드를 이해할때 더 쉬운 형태에 대해서 고민해보시면 좋을것 같아요 😄

csh0034 reviewed (changes requested):
리뷰 반영 잘해주셨네요 😄 
몇가지 코멘트 추가로 남겨두었는데 
확인 부탁드립니다!

ksh9891 committed:
fix(lms): 필드명 수정

ksh9891 committed:
fix(lms): 패키지 구조 변경

ksh9891 committed:
fix(lms): 예외를 RuntimeException 하위 표준 예외로 변경

ksh9891 committed:
fix(lms): 예외를 RuntimeException 하위 표준 예외로 변경

ksh9891 committed:
fix(lms): Session 이 SessionUsers 를 가지도록 수정

ksh9891 commented:
안녕하세요. 리뷰어님
피드백 주신부분 반영하여 다시 리뷰 요청드립니다.
생각보다 구조를 잡기가 쉽지가 않은 것 같네요 ㅠㅠ
이번에도 잘 부탁드립니다.

csh0034 commented on src/main/java/nextstep/session/domain/SessionUser.java:
```diff
@@ -0,0 +1,28 @@
+public class SessionUser {
+    private final NsUser user;
+    private final Session session;
+    private final Payment payment;
```
Payment 에서 sessionId nsUserId 를 모두 관리하고 있는데 

SessionUser 에서도 관리가 필요한걸까요? 🤔

ksh9891 replied:
고민을 해보았는데요. SessionUser 는 Payment 를 몰라도 될 것 같습니다.
SessionUser  는 참여신청한 유저와 세션을 이어주는 역할만하지 "해당 사용자가 결제를 했는가?", "하지 않았는가?" 는 중요치 않은 것 같습니다.

csh0034 reviewed (approved):
수고 많으셨습니다 👍 
다음단계에서 봬요 😄
